### PR TITLE
Better use of column index

### PR DESCRIPTION
**Pull request by generall**
### All Submissions:

related to https://github.com/qdrant/qdrant/issues/460

This PR is a continuation of the work for enabling column-based payload index in the search process.
On the previous step it was only possible to use primary condition as a fail-fast strategy for condition checking, but actual filtering queries may contain multiple fields 

### Main optimization hypothesizes:

* it is better to avoid using payload storage during the search as much as possible. Payload uploading will be even slower when we enable on-disk payload storage
* If worth spending some time optimizing the query on the beginning, cause query is usually evaluated thousand times during the search

### Approach

We introduce a new Structure, called `OptimizedFilter`. It is similar to regular filter, but all conditions are replaces with closure `ConditionCheckerFn`. This closure is obtained from the original condition and holds all required information to perform the condition check for a specified `PointOffsetId`. Optimized filter is generated from the original filter during the construction of filtering context. It is equivalent to the original filter is equivalent exactly to the order of execution - more valuable conditions are executed first and that is how `PrimaryCondition` logic is generalized.

### Main challenges:

* Rust lifetimes - Filtering context is created during the query, so it should avoid copying as much as possible. It relies on references to the structures inside the payload index which makes things complicated
* Borrow checker - Same as above
* Traits - We are going to have custom implementation of condition checking which depends on some functions, which are only implemented in some indexes (not available in plain storage/index), so we will need to know exact type of payload storage during the check execution. In this PR it is solved with Enum DIspatch

### Results

Screenshot from 2022-04-16 02-16-19

----

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you lint your code locally using ``cargo fmt`` command prior to submission?
3. [x] Have you checked your code using ```cargo clippy``` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/segment/src/index/query_optimization/optimized_filter.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+pub type ConditionCheckerFn<'a> = Box<dyn Fn(PointOffsetType) -> bool + 'a>;
+
+pub enum OptimizedCondition<'a> {
+    Checker(ConditionCheckerFn<'a>),
+    /// Nested filter
+    Filter(OptimizedFilter<'a>),
+}
+
+pub struct OptimizedFilter<'a> {
+    /// At least one of thous conditions should match
```
Comment (agourlay) [inline lib/segment/src/index/query_optimization/optimized_filter.rs:12]:
```suggestion
    /// At least one of those conditions should match
```

File: lib/segment/src/payload_storage/query_checker.rs [hunk 2]
```diff
@@ -62,79 +63,84 @@ where
...
+            Value::Array(array) => array.is_empty(),
+            _ => false,
+        },
+    }
+}
+
+pub fn check_field_condition(field_condition: &FieldCondition, payload: &Payload) -> bool {
+    payload.get_value(&field_condition.key).map_or(false, |p| {
+        let mut res = false;
+        // ToDo: Convert onto iterator over checkers, so it would be impossible to forget a condition
```
Comment (agourlay) [inline lib/segment/src/payload_storage/query_checker.rs:105]: Would be nice to implement this ToDo :)

**Review by agourlay — COMMENTED** [review]
Performed a first pass to understand what's happening here and drop a few comments.
Will do another pass later today.

File: lib/segment/src/payload_storage/query_checker.rs (hunk 2 above)
Comment (generall) [reply lib/segment/src/payload_storage/query_checker.rs:105]: the thing is, i don't know how.

**Review by gvelo — APPROVED** [review]

**Review by agourlay — APPROVED** [review]

**Pull request merged**